{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[92mMinimum number of gene clusters = \n",
      "33\n",
      "Maximum number of gene clusters = \n",
      "81\n",
      "\u001b[0m\u001b[92mFound neutral solutions = \n",
      "5\n",
      "\u001b[0m\u001b[92mFound clonal solutions = \n",
      "{}\n",
      "\u001b[0m\u001b[92mFound scales = \n",
      "()\n",
      "\u001b[0m\u001b[92mMaximum copy-number = \n",
      "8.0\n",
      "\u001b[0m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "# ------------------------------------------\n",
      "#   Problem Information\n",
      "# ------------------------------------------\n",
      "#     Number of constraints: 7513\n",
      "#     Number of variables: 3064\n",
      "# ------------------------------------------\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[92mMaximum copy-number = \n",
      "8.0\n",
      "\u001b[0m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "# ------------------------------------------\n",
      "#   Problem Information\n",
      "# ------------------------------------------\n",
      "#     Number of constraints: 10884\n",
      "#     Number of variables: 4312\n",
      "# ------------------------------------------\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[92mMaximum copy-number = \n",
      "8.0\n",
      "\u001b[0m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "# ------------------------------------------\n",
      "#   Problem Information\n",
      "# ------------------------------------------\n",
      "#     Number of constraints: 14255\n",
      "#     Number of variables: 5560\n",
      "# ------------------------------------------\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[92mMaximum copy-number = \n",
      "8.0\n",
      "\u001b[0m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "# ------------------------------------------\n",
      "#   Problem Information\n",
      "# ------------------------------------------\n",
      "#     Number of constraints: 17626\n",
      "#     Number of variables: 6808\n",
      "# ------------------------------------------\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[92mMaximum copy-number = \n",
      "8.0\n",
      "\u001b[0m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "# ------------------------------------------\n",
      "#   Problem Information\n",
      "# ------------------------------------------\n",
      "#     Number of constraints: 20997\n",
      "#     Number of variables: 8056\n",
      "# ------------------------------------------\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[92mMaximum copy-number = \n",
      "8.0\n",
      "\u001b[0m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "# ------------------------------------------\n",
      "#   Problem Information\n",
      "# ------------------------------------------\n",
      "#     Number of constraints: 24368\n",
      "#     Number of variables: 9304\n",
      "# ------------------------------------------\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[92mMaximum copy-number = \n",
      "8.0\n",
      "\u001b[0m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "# ------------------------------------------\n",
      "#   Problem Information\n",
      "# ------------------------------------------\n",
      "#     Number of constraints: 27739\n",
      "#     Number of variables: 10552\n",
      "# ------------------------------------------\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[92mDiploid objective values = \n",
      "{2: 202.58886988227158, 3: 111.02099570964599, 4: 93.93223628125641, 5: 76.28063190364306, 6: 64.9247487110025, 7: 58.38926401159371, 8: 52.69431953399958}\n",
      "\u001b[0m"
     ]
    },
    {
     "ename": "IndexError",
     "evalue": "tuple index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mcompute_cn_pcf\u001b[39;00m \u001b[39mimport\u001b[39;00m compute_cn_pcf\n\u001b[0;32m----> 2\u001b[0m dbest,tbest\u001b[39m=\u001b[39mcompute_cn_pcf(\n\u001b[1;32m      3\u001b[0m     \u001b[39m\"\u001b[39;49m\u001b[39mTR273/TR273.fx\u001b[39;49m\u001b[39m\"\u001b[39;49m,\n\u001b[1;32m      4\u001b[0m     maxC\u001b[39m=\u001b[39;49m\u001b[39m8\u001b[39;49m,\n\u001b[1;32m      5\u001b[0m     tS\u001b[39m=\u001b[39;49m\u001b[39m0.75\u001b[39;49m,\n\u001b[1;32m      6\u001b[0m     preserve_arm\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m      7\u001b[0m     ampdel\u001b[39m=\u001b[39;49m\u001b[39mTrue\u001b[39;49;00m,\n\u001b[1;32m      8\u001b[0m     n_seed\u001b[39m=\u001b[39;49m\u001b[39m20\u001b[39;49m,\n\u001b[1;32m      9\u001b[0m     rundir\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39mTR273\u001b[39;49m\u001b[39m\"\u001b[39;49m\n\u001b[1;32m     10\u001b[0m     )\n",
      "File \u001b[0;32m~/HATCHet_PCF/src/compute_cn_pcf.py:125\u001b[0m, in \u001b[0;36mcompute_cn_pcf\u001b[0;34m(bbfile, minK, maxK, covar, decode_alg, tmat, tau, restarts, preserve_arm, subset, tD, tB, tR, tS, minC, maxC, solver, solve_mode, d, cn_max, mu, dT, g, limit, ampdel, n_seed, n_worker, random_seed, max_iters, timelimit, rundir)\u001b[0m\n\u001b[1;32m     98\u001b[0m diploid\u001b[39m=\u001b[39mrunningDiploid(\n\u001b[1;32m     99\u001b[0m     df\u001b[39m=\u001b[39mseg,\n\u001b[1;32m    100\u001b[0m     neutral\u001b[39m=\u001b[39mneutral,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    114\u001b[0m     timelimit\u001b[39m=\u001b[39mtimelimit,\n\u001b[1;32m    115\u001b[0m     rundir\u001b[39m=\u001b[39mrundir)\n\u001b[1;32m    119\u001b[0m sp\u001b[39m.\u001b[39mlog(\n\u001b[1;32m    120\u001b[0m     msg\u001b[39m=\u001b[39m\u001b[39m'\u001b[39m\u001b[39mDiploid objective values = \u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39m'\u001b[39m \u001b[39m+\u001b[39m \u001b[39mstr\u001b[39m(diploid) \u001b[39m+\u001b[39m \u001b[39m'\u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39m'\u001b[39m,\n\u001b[1;32m    121\u001b[0m     level\u001b[39m=\u001b[39m\u001b[39m'\u001b[39m\u001b[39mINFO\u001b[39m\u001b[39m'\u001b[39m,\n\u001b[1;32m    122\u001b[0m )\n\u001b[0;32m--> 125\u001b[0m tetraploid\u001b[39m=\u001b[39mrunningTetraploid(\n\u001b[1;32m    126\u001b[0m     df\u001b[39m=\u001b[39;49mseg,\n\u001b[1;32m    127\u001b[0m     clonal\u001b[39m=\u001b[39;49mclonal,\n\u001b[1;32m    128\u001b[0m     scale\u001b[39m=\u001b[39;49mscale,\n\u001b[1;32m    129\u001b[0m     minC\u001b[39m=\u001b[39;49mminC, \n\u001b[1;32m    130\u001b[0m     maxC\u001b[39m=\u001b[39;49mmaxC, \n\u001b[1;32m    131\u001b[0m     solver\u001b[39m=\u001b[39;49msolver,\n\u001b[1;32m    132\u001b[0m     solve_mode\u001b[39m=\u001b[39;49msolve_mode,\n\u001b[1;32m    133\u001b[0m     d\u001b[39m=\u001b[39;49md,\n\u001b[1;32m    134\u001b[0m     cn_max\u001b[39m=\u001b[39;49mcn_max,\n\u001b[1;32m    135\u001b[0m     mu\u001b[39m=\u001b[39;49mmu,\n\u001b[1;32m    136\u001b[0m     dT\u001b[39m=\u001b[39;49mdT,\n\u001b[1;32m    137\u001b[0m     ampdel\u001b[39m=\u001b[39;49mampdel,\n\u001b[1;32m    138\u001b[0m     n_seed\u001b[39m=\u001b[39;49mn_seed,\n\u001b[1;32m    139\u001b[0m     n_worker\u001b[39m=\u001b[39;49mn_worker,\n\u001b[1;32m    140\u001b[0m     random_seed\u001b[39m=\u001b[39;49mrandom_seed,\n\u001b[1;32m    141\u001b[0m     max_iters\u001b[39m=\u001b[39;49mmax_iters,\n\u001b[1;32m    142\u001b[0m     timelimit\u001b[39m=\u001b[39;49mtimelimit,\n\u001b[1;32m    143\u001b[0m     rundir\u001b[39m=\u001b[39;49mrundir\n\u001b[1;32m    144\u001b[0m )\n\u001b[1;32m    148\u001b[0m sp\u001b[39m.\u001b[39mlog(\n\u001b[1;32m    149\u001b[0m     msg\u001b[39m=\u001b[39m\u001b[39m'\u001b[39m\u001b[39mTetraploid objective values = \u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39m'\u001b[39m \u001b[39m+\u001b[39m \u001b[39mstr\u001b[39m(tetraploid) \u001b[39m+\u001b[39m \u001b[39m'\u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39m'\u001b[39m,\n\u001b[1;32m    150\u001b[0m     level\u001b[39m=\u001b[39m\u001b[39m'\u001b[39m\u001b[39mINFO\u001b[39m\u001b[39m'\u001b[39m,\n\u001b[1;32m    151\u001b[0m )\n\u001b[1;32m    153\u001b[0m dbest,tbest\u001b[39m=\u001b[39mselect(\n\u001b[1;32m    154\u001b[0m     diploid\u001b[39m=\u001b[39mdiploid,\n\u001b[1;32m    155\u001b[0m     tetraploid\u001b[39m=\u001b[39mtetraploid,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    158\u001b[0m     rundir\u001b[39m=\u001b[39mrundir\n\u001b[1;32m    159\u001b[0m )\n",
      "File \u001b[0;32m~/HATCHet_PCF/src/execute_solver_pcf.py:120\u001b[0m, in \u001b[0;36mrunningTetraploid\u001b[0;34m(df, clonal, scale, minC, maxC, solver, solve_mode, d, cn_max, mu, dT, ampdel, n_seed, n_worker, random_seed, max_iters, timelimit, rundir)\u001b[0m\n\u001b[1;32m     99\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mrunningTetraploid\u001b[39m(\n\u001b[1;32m    100\u001b[0m     df,\n\u001b[1;32m    101\u001b[0m     clonal,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    117\u001b[0m     rundir\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    118\u001b[0m ):\n\u001b[0;32m--> 120\u001b[0m     cn \u001b[39m=\u001b[39m \u001b[39m'\u001b[39m\u001b[39m{}\u001b[39;00m\u001b[39m:\u001b[39m\u001b[39m{}\u001b[39;00m\u001b[39m:\u001b[39m\u001b[39m{}\u001b[39;00m\u001b[39m'\u001b[39m\u001b[39m.\u001b[39mformat(scale[\u001b[39m0\u001b[39;49m], clonal[scale[\u001b[39m0\u001b[39m]][\u001b[39m0\u001b[39m], clonal[scale[\u001b[39m0\u001b[39m]][\u001b[39m1\u001b[39m])\n\u001b[1;32m    121\u001b[0m     cn \u001b[39m+\u001b[39m\u001b[39m=\u001b[39m \u001b[39m'\u001b[39m\u001b[39m,\u001b[39m\u001b[39m{}\u001b[39;00m\u001b[39m:\u001b[39m\u001b[39m{}\u001b[39;00m\u001b[39m:\u001b[39m\u001b[39m{}\u001b[39;00m\u001b[39m'\u001b[39m\u001b[39m.\u001b[39mformat(scale[\u001b[39m1\u001b[39m], clonal[scale[\u001b[39m1\u001b[39m]][\u001b[39m0\u001b[39m], clonal[scale[\u001b[39m1\u001b[39m]][\u001b[39m1\u001b[39m])\n\u001b[1;32m    122\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mlen\u001b[39m(clonal) \u001b[39m>\u001b[39m \u001b[39m2\u001b[39m:\n",
      "\u001b[0;31mIndexError\u001b[0m: tuple index out of range"
     ]
    }
   ],
   "source": [
    "from compute_cn_pcf import compute_cn_pcf\n",
    "import argparse\n",
    "\n",
    "def arg_parse(args=None):\n",
    "    \"\"\"\n",
    "    Parse command line arguments\n",
    "    Returns:\n",
    "    \"\"\"\n",
    "    description = ''\n",
    "    parser = argparse.ArgumentParser(\n",
    "        prog='hatchet_pcf compute-cn',\n",
    "        description=description,\n",
    "        formatter_class=argparse.RawTextHelpFormatter,\n",
    "    )\n",
    "    parser.add_argument('SOLVER', type=str, nargs='?', help='Path to the executable solver')\n",
    "    parser.add_argument(\n",
    "        '-i',\n",
    "        '--bbfile',\n",
    "        type=str,\n",
    "        required=True,\n",
    "        help='Prefix path to bb file (required)',\n",
    "    )\n",
    "\n",
    "    parser.add_argument(\n",
    "        '-m',\n",
    "        '--minK',\n",
    "        type=int,\n",
    "        required=True,\n",
    "        default=None,\n",
    "        help='Minimum number of cluster to infer. Default (None: Number of chromosomal arms analyzed)',\n",
    "    )\n",
    "\n",
    "    parser.add_argument(\n",
    "        '-M',\n",
    "        '--maxK',\n",
    "        type=int,\n",
    "        required=True,\n",
    "        default=None,\n",
    "        help='Maximum number of clones. Default (None: Number of genes used as input)',\n",
    "    )\n",
    "    \n",
    "    parser.add_argument(\n",
    "        '-v',\n",
    "        '--covar',\n",
    "        type=str,\n",
    "        required=True,\n",
    "        map='diag',\n",
    "        help='Form of covariance matrix: spherical, diag, full, or tied (default: diag)',\n",
    "    )\n",
    "\n",
    "    parser.add_argument(\n",
    "        '-d',\n",
    "        '--decode_alg',\n",
    "        type=str,\n",
    "        required=True,\n",
    "        default='map',\n",
    "        help='Decoding algorithm to use: map or viterbi (default: map)',\n",
    "    )\n",
    "\n",
    "    parser.add_argument(\n",
    "        '-t',\n",
    "        '--tmat',\n",
    "        type=str,\n",
    "        required=True,\n",
    "        default='diag',\n",
    "        help='Form of transition matrix to infer: fixed, diag (1-parameter), or full (default: diag)',\n",
    "    )\n",
    "\n",
    "\n",
    "    parser.add_argument(\n",
    "        '-T',\n",
    "        '--tau',\n",
    "        type=float,\n",
    "        required=True,\n",
    "        default=0.000001,\n",
    "        help='Off-diagonal value for initializing transition matrix (Default: 0.000001)'\n",
    "    )\n",
    "\n",
    "    parser.add_argument(\n",
    "        '-r',\n",
    "        '--restarts',\n",
    "        type=int,\n",
    "        required=True,\n",
    "        default=10,\n",
    "        help=\"Number of restarts performed by the clustering to choose the best (default: 10)\",\n",
    "    )\n",
    "\n",
    "\n",
    "    parser.add_argument(\n",
    "        '-y',\n",
    "        '--subset',\n",
    "        required=False,\n",
    "        default=None,\n",
    "        type=str,\n",
    "        nargs='+',\n",
    "        help=(\n",
    "            'List of sample names to use as a subset of those included in binning' '(default: none, run on all samples)'\n",
    "        ),\n",
    "    )\n",
    "\n",
    "\n",
    "    parser.add_argument(\n",
    "        '-p',\n",
    "        '--preserve_arm',\n",
    "        required=False,\n",
    "        default=False,\n",
    "        type=bool,\n",
    "        help='Preserve chromosome arms in local clustering (default: none)'\n",
    "    )\n",
    "\n",
    "\n",
    "    parser.add_argument(\n",
    "        '-x',\n",
    "        '--runningdir',\n",
    "        type=str,\n",
    "        required=False,\n",
    "        default=\"./\",\n",
    "        help='Running directory (default: ./)',\n",
    "    )\n",
    "   \n",
    "    parser.add_argument(\n",
    "        '-f',\n",
    "        '--noampdel',\n",
    "        action='store_true',\n",
    "        default=True,\n",
    "        required=False,\n",
    "        help=\n",
    "            'Remove amp-del assumption where each mutated allele of every segment can be either amplified or '\n",
    "            'deleted in all tumor clones w.r.t. base (2 for diploid and 4 for tetraploid) (default: use assumption)'\n",
    "        ,\n",
    "    )\n",
    "\n",
    "    parser.add_argument(\n",
    "        '-d',\n",
    "        '--cnstates',\n",
    "        type=int,\n",
    "        required=False,\n",
    "        default=-1,\n",
    "        help='Maximum number of distinct copy-number states for each segment (default: None, no limit)',\n",
    "    )\n",
    "\n",
    "    parser.add_argument(\n",
    "        '-tS',\n",
    "        type=float,\n",
    "        required=False,\n",
    "        default=0.75,\n",
    "        help='The minimum proportion of samples within clonal cluster to call cluster as clonal (default: 0.75)',\n",
    "    )\n",
    "\n",
    "    parser.add_argument(\n",
    "        '-tR',\n",
    "        type=float,\n",
    "        required=False,\n",
    "        default=0.08,\n",
    "        help='RDR tolerance for clonal clusters (default: 0.08)',\n",
    "    )\n",
    "\n",
    "    parser.add_argument(\n",
    "        '-tB',\n",
    "        type=float,\n",
    "        required=False,\n",
    "        default=0.04,\n",
    "        help='BAF tolerance for clonal clusters (default: 0.04)',\n",
    "    )\n",
    "\n",
    "    parser.add_argument(\n",
    "        '-tD',\n",
    "        type=float,\n",
    "        required=False,\n",
    "        default=0.1,\n",
    "        help=(\n",
    "            'Maximum BAF shift for neutral cluster used to automatically infer the diploid/tetraploid cluster '\n",
    "            '(default: 0.1)'\n",
    "        ),\n",
    "    )\n",
    "\n",
    "    parser.add_argument(\n",
    "        '-minC',\n",
    "        type=int,\n",
    "        required=True,\n",
    "        default=2,\n",
    "        help='Minimum number of clines. default: 2',\n",
    "    )\n",
    "\n",
    "    parser.add_argument(\n",
    "        '-maxC',\n",
    "        type=int,\n",
    "        required=True,\n",
    "        default=8,\n",
    "        help='Maximum number of clones. default: 8',\n",
    "    )\n",
    "\n",
    "\n",
    "    parser.add_argument(\n",
    "        '-solver',\n",
    "        type=str,\n",
    "        required=True,\n",
    "        default=\"cpp\",\n",
    "        help='Solver type to use. default cpp',\n",
    "    )\n",
    "\n",
    "    parser.add_argument(\n",
    "        '-solvermode',\n",
    "        type=str,\n",
    "        required=True,\n",
    "        default=\"cd\",\n",
    "        help='Solver type to use. default cd',\n",
    "    )\n",
    "\n",
    "    parser.add_argument(\n",
    "        '-cnmax',\n",
    "        type=int,\n",
    "        required=True,\n",
    "        default=None,\n",
    "        help='Maximum number of copy number. default None',\n",
    "    )\n",
    "\n",
    "\n",
    "    parser.add_argument(\n",
    "        '-l',\n",
    "        '--limitinc',\n",
    "        type=float,\n",
    "        required=False,\n",
    "        default=0.6,\n",
    "        help=(\n",
    "            'Upper bound to the relative increase of objective function. When there are significant small CNAs, '\n",
    "            'their effect on the objective function may be confounded by only larger events, use this value to '\n",
    "            'limit the relative increase of OBJ so that fitting small CNAs is more considered (default: None)'\n",
    "        ),\n",
    "    )\n",
    "\n",
    "    parser.add_argument(\n",
    "        '-g',\n",
    "        '--ghostprop',\n",
    "        type=float,\n",
    "        required=False,\n",
    "        default=0.3,\n",
    "        help=(\n",
    "            'Increasing proportion used to compute the value of the first ghost point added in the solution '\n",
    "            'selection (default: 0.3)'\n",
    "        ),\n",
    "    )\n",
    "\n",
    "    parser.add_argument(\n",
    "        '-u',\n",
    "        '--mu',\n",
    "        type=float,\n",
    "        required=False,\n",
    "        default=0.03,\n",
    "        help='Minimum clone proporion in each sample (default: 0.03)',\n",
    "    )\n",
    "\n",
    "    parser.add_argument(\n",
    "        '--maxiterations',\n",
    "        type=int,\n",
    "        required=False,\n",
    "        default=10,\n",
    "        help='Maximum number of iterations composed of C-step/U-step for each seed (default: 10)',\n",
    "    )\n",
    "\n",
    "    parser.add_argument(\n",
    "        '-p',\n",
    "        '--seeds',\n",
    "        type=int,\n",
    "        required=False,\n",
    "        default=400,\n",
    "        help='Number of seeds for coordinate-descent method (default: 400)',\n",
    "    )\n",
    "\n",
    "    parser.add_argument(\n",
    "        '-j',\n",
    "        '--jobs',\n",
    "        type=int,\n",
    "        required=False,\n",
    "        default=None,\n",
    "        help='Number of parallel jobs (default: maximum available on the machine)',\n",
    "    )\n",
    "\n",
    "    parser.add_argument(\n",
    "        '-r',\n",
    "        '--randomseed',\n",
    "        type=int,\n",
    "        required=False,\n",
    "        default=None,\n",
    "        help='Random seed (default: None)',\n",
    "    )\n",
    "    parser.add_argument(\n",
    "        '-s',\n",
    "        '--timelimit',\n",
    "        type=int,\n",
    "        required=False,\n",
    "        default=None,\n",
    "        help='Time limit for each ILP run (default: None)',\n",
    "    )\n",
    "    parser.add_argument(\n",
    "        '-m',\n",
    "        '--memlimit',\n",
    "        type=int,\n",
    "        required=False,\n",
    "        default=None,\n",
    "        help='Memory limit for each ILP run (default: None)',\n",
    "    )\n",
    "\n",
    "    \n",
    "    args = parser.parse_args(args)\n",
    "    return args\n",
    "\n",
    "\n",
    "\n",
    "dbest,tbest=compute_cn_pcf(\n",
    "    \"TR273/TR273.fx\",\n",
    "    maxC=8,\n",
    "    tS=0.75,\n",
    "    preserve_arm=False,\n",
    "    ampdel=True,\n",
    "    n_seed=20,\n",
    "    rundir=\"TR273\"\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "usage: hatchet_pcf compute-cn [-h] [-i BBFILE] [-m MINK] [-M MAXK] [-v COVAR]\n",
      "                              [-z DECODE_ALG] [-t TMAT] [-T TAU] [-r RESTARTS]\n",
      "                              [-y SUBSET [SUBSET ...]] [-p PRESERVE_ARM]\n",
      "                              [-x RUNNINGDIR] [-f] [-d CNSTATES] [-tS TS]\n",
      "                              [-tR TR] [-tB TB] [-tD TD] [-minC MINC]\n",
      "                              [-maxC MAXC] [-solver SOLVER]\n",
      "                              [-solvermode SOLVERMODE] [-cnmax CNMAX]\n",
      "                              [-l LIMITINC] [-g GHOSTPROP] [-u MU]\n",
      "                              [--maxiterations MAXITERATIONS] [-s SEEDS]\n",
      "                              [-j JOBS] [-rs RANDOMSEED] [-tl TIMELIMIT]\n",
      "                              [-ml MEMLIMIT]\n",
      "hatchet_pcf compute-cn: error: unrecognized arguments: --ip=127.0.0.1 --stdin=9003 --control=9001 --hb=9000 --Session.signature_scheme=\"hmac-sha256\" --Session.key=b\"e4e1ec41-aea5-4fb9-a476-3afaa445a470\" --shell=9002 --transport=\"tcp\" --iopub=9004 --f=/home/vscode/.local/share/jupyter/runtime/kernel-v2-402W9wBYAv0yvDt.json\n"
     ]
    },
    {
     "ename": "SystemExit",
     "evalue": "2",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[0;31mSystemExit\u001b[0m\u001b[0;31m:\u001b[0m 2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/vscode/.local/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3516: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.\n",
      "  warn(\"To exit: use 'exit', 'quit', or Ctrl-D.\", stacklevel=1)\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.2 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
